docs/design/operator.md

```diff
@@ -0,0 +1,200 @@
+* **Healthy:** Whether or not all processes are running correctly.
+* **Replicas:** How many replicas of this node definition to deploy.
+* **Devices:** A list of device definitions.
+
+Finally, the operator will maintain the following state parameters for a given
+device:
+
+* **Name:** An identifier that is unique across the entire cluster.
+* **Provision:** Whether or not to provision the PersistentVolume backing the
+  device.
+
+## Features
+
+Since the Gluster operator will support a variety of deployment configurations,
+it will not be feasible to provide identical feature sets across all
+configurations. The following sections note which features will be available to
+which configuration.
+
+All operational features will be designed to be as idempotent as possible. It is
+acknowledged that this may not be immediately possible due to restrictions in
+the underlying technology but an effort will still be made.
+
+### Deployment
+
+The Gluster operator will be able to deploy a new Gluster cluster using a set of
+configuration parameters. This configuration will allow for extensive
+customization if desired but will assume sane defaults for as many parameters as
+possible to allow for as minimal a configuration as possible.
+
+**Target Configurations:** External, Hosted, Managed
+
+### Cluster Upgrade
+
+The Gluster operator will be able to execute a rolling upgrade of any hosted or
+managed Gluster cluster. It will trigger an upgrade one node at a time, waiting
+for the node to come up and for the cluster to report no pending heals to any
+files before triggering a subsequent node.
+
+**Target Configurations:** Hosted, Managed
+
+### Event Monitoring
+
+The Gluster operator will watch for events on various aspects of the clusters it
+will automate.
+
+#### Storage Capacity
+
+The Gluster operator will monitor the storage capacity of the overall cluster.
+Storage capacity is defined here as the amount of space that can be committed to
+new Gluster volumes and not how much free space is available. It will trigger
+events when storage capacity reaches "Warning" levels and "Critical" levels.
+These levels will be configurable, and should be set to a value that will allow
+for corrective action to take place before the storage reaches maximum capacity.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Volume Density
+
+The Gluster operator will monitor the total volume density of any cluster it is
+automating. It will trigger events when the number of volumes in the cluster
+reaches a pre-defined level that should prompt corrective action. This value
+will be configurable, and should be set to a value that will allow for
+corrective action to take place before cluster nodes lag or crash.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Device Health
+
+The Gluster operator will monitor the health of all devices in an automated
+Gluster cluster. It will report a "Warning" when a device reports failing health
```

Comment:
I think we need to define the term "automated Gluster cluster" in this document. (see below comments)